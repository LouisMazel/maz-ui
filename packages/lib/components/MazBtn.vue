<template>
  <Component
    :is="component"
    :disabled="isDisabled"
    class="m-btn"
    :class="[
      `--${size}`,
      btnColorClass,
      cursorClass,
      variantClass,
      {
        '--block': block,
        '--no-underline': noUnderline,
        '--no-leading': noLeading,
        '--fab': fab,
        '--loading': loading,
        '--disabled': isDisabled,
        '--icon': hasIcon,
        '--rounded': rounded,
        '--no-rounded': noRounded,
        '--no-padding': noPadding,
        '--no-elevation': noElevation,
      },
    ]"
    :type="btnType"
  >
    <div
      v-if="hasLeftIcon"
      class="m-btn__icon-left maz-flex maz-flex-center"
      :class="{ 'maz-invisible': hasLoader }"
    >
      <!--
        @slot left-icon - The icon to display on the left of the button
      -->
      <slot name="left-icon">
        <MazIcon v-if="typeof leftIcon === 'string'" :name="leftIcon" />
        <Component :is="leftIcon" v-else-if="leftIcon" />
      </slot>
    </div>

    <div v-if="hasFabIcon" class="m-btn__icon" :class="{ 'maz-invisible': hasLoader }">
      <!--
        @slot icon - The icon to display on the fab button
      -->
      <slot name="icon">
        <MazIcon v-if="typeof icon === 'string'" :name="icon" />
        <Component :is="icon" v-else-if="icon" />
      </slot>
    </div>

    <span v-if="$slots.default" :class="[{ 'maz-invisible': hasLoader }, contentClass]">
      <!--
        @slot default - The content of the button
      -->
      <slot></slot>
    </span>

    <div v-if="hasRightIcon" class="m-btn__icon-right" :class="{ 'maz-invisible': hasLoader }">
      <!--
        @slot right-icon - The icon to display on the right of the button
      -->
      <slot name="right-icon">
        <MazIcon v-if="typeof rightIcon === 'string'" :name="rightIcon" />
        <Component :is="rightIcon" v-else-if="rightIcon" />
      </slot>
    </div>

    <MazSpinner v-if="hasLoader" class="m-btn-loader" size="2em" :color="color" />
  </Component>
</template>

<script lang="ts" setup>
  import {
    computed,
    useAttrs,
    useSlots,
    defineAsyncComponent,
    onBeforeMount,
    type FunctionalComponent,
    type SVGAttributes,
    type ComponentPublicInstance,
    type Component,
  } from 'vue'

  import type { Color, Size } from './types'
  export type { Color, Size }

  const MazSpinner = defineAsyncComponent(() => import('./MazSpinner.vue'))
  const MazIcon = defineAsyncComponent(() => import('./MazIcon.vue'))

  const { href, to } = useAttrs()
  const slots = useSlots()

  type Icon = FunctionalComponent<SVGAttributes> | ComponentPublicInstance | Component

  const props = withDefaults(
    defineProps<{
      /** The variant of the button - Change UI of component (link or button style)
       * @values `'button' | 'link'`
       * */
      variant?: 'button' | 'link'
      /**
       * The size of the button
       * @values `'xl' | 'lg' | 'md' | 'sm' | 'xs' | 'mini'`
       */
      size?: Size
      /**
       * The color of the button
       * @values `'primary' | 'secondary' | 'info' | 'success' | 'warning' | 'danger' | 'white' | 'black' | 'transparent' | 'theme'`
       */
      color?: Color
      /**
       * The type of the button
       * @values `'submit' | 'reset' | 'button'`
       */
      type?: 'submit' | 'reset' | 'button'
      /**
       * If true, the button will have a full border radius
       * @default false
       */
      rounded?: boolean
      /**
       * If true, the button will have no border radius
       * @default false
       */
      noRounded?: boolean
      /**
       * If true, the button have the "border" style
       * @default false
       */
      outline?: boolean
      /**
       * If true, the button will have a pastel color
       * @default false
       */
      pastel?: boolean
      /**
       * If true, the button will have a full width
       * @default false
       */
      block?: boolean
      /** If true, the button will have no underline on hover (useful with `variant="link"`)
       * @default false
       */
      noUnderline?: boolean
      /**
       * If true, the button will have no leading (useful with `variant="link"`)
       * @default false
       */
      noLeading?: boolean
      /**
       * Enable the button loader
       * @default false
       */
      loading?: boolean
      /**
       * Disable the button
       * @default false
       */
      disabled?: boolean
      /**
       * If true, the button will have a fab style
       * @default false
       */
      fab?: boolean
      /**
       * The name of the icon to display or component, only with fab
       * `@type` `{string | FunctionalComponent<SVGAttributes> | ComponentPublicInstance | Component}`
       */
      icon?: string | Icon
      /**
       * The name of the icon or component to display on the left of the button
       * `@type` `{string | FunctionalComponent<SVGAttributes> | ComponentPublicInstance | Component}`
       */
      leftIcon?: string | Icon
      /**
       * The name of the icon or component to display on the right of the button
       * `@type` `{string | FunctionalComponent<SVGAttributes> | ComponentPublicInstance | Component}`
       */
      rightIcon?: string | Icon
      /**
       * If true, the button will have no padding
       * @default false
       */
      noPadding?: boolean
      /**
       * If true, the button will have no box-shadow
       * @default false
       */
      noElevation?: boolean
      /** The class applied to the content wrapper (<span />) of the button */
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      contentClass?: any
    }>(),
    {
      variant: 'button',
      size: 'md',
      color: 'primary',
      type: 'button',
      icon: undefined,
      leftIcon: undefined,
      rightIcon: undefined,
      contentClass: undefined,
    },
  )

  onBeforeMount(() => {
    if (props.icon && !props.fab) {
      console.error('[maz-ui](MazBtn) the prop "icon" must be used only with "fab" props')
    }
  })

  const component = computed(() => {
    if (href) return 'a'
    else if (to) return 'router-link'
    return 'button'
  })

  const btnColorClass = computed(() =>
    props.pastel
      ? `--${props.color}-pastel`
      : props.outline
        ? `--${props.color}-outline`
        : `--${props.color}`,
  )
  const isDisabled = computed(
    () => (props.loading || props.disabled) && component.value === 'button',
  )
  const cursorClass = computed(() => (isDisabled.value ? '--cursor-default' : '--cursor-pointer'))
  const variantClass = computed(() => `--is-${props.variant}`)
  const hasLoader = computed(() => props.loading && props.variant === 'button')
  const hasLeftIcon = computed(() => !!slots['left-icon'] || props.leftIcon)
  const hasRightIcon = computed(() => !!slots['right-icon'] || props.rightIcon)
  const hasIcon = computed(() => hasLeftIcon.value || hasRightIcon.value)
  const hasFabIcon = computed(() => props.fab && (props.icon || !!slots['icon']))
  const btnType = computed(() => (component.value === 'button' ? props.type : undefined))
</script>

<style lang="postcss" scoped>
  .m-btn {
    @apply maz-flex maz-flex-none maz-items-center maz-gap-2 maz-border maz-border-solid maz-border-transparent maz-text-center maz-text-base maz-text-normal;

    & span {
      @apply maz-leading-none;
    }

    &-loader {
      @apply maz-absolute;
    }

    &.--cursor-pointer {
      @apply maz-cursor-pointer;
    }

    &.--cursor-default {
      @apply maz-cursor-default;
    }

    &.--is-link {
      @apply maz-inline-flex maz-items-center maz-bg-transparent
        maz-outline-none maz-transition maz-duration-200 maz-ease-in-out;

      &:not(.--no-leading) span {
        @apply maz-leading-9;
      }

      &:not(:disabled):hover,
      &:not(:disabled):focus {
        &:not(.--no-underline) {
          @apply maz-underline;
        }
      }

      &.--secondary {
        @apply maz-text-secondary;
      }

      &.--info {
        @apply maz-text-info;
      }

      &.--warning {
        @apply maz-text-warning-600;
      }

      &.--danger {
        @apply maz-text-danger-600;
      }

      &.--success {
        @apply maz-text-success-600;
      }

      &.--white {
        @apply maz-text-white;
      }

      &.--black {
        @apply maz-text-black;
      }

      &.--theme {
        @apply maz-text-normal;
      }
    }

    &.--is-button {
      @apply maz-relative maz-inline-flex maz-items-center maz-justify-center maz-overflow-hidden
        maz-border-transparent maz-bg-transparent maz-font-medium maz-no-underline
        maz-transition maz-duration-300 maz-ease-in-out;

      &:not(.--no-rounded) {
        @apply maz-rounded;
      }

      &.--rounded {
        @apply maz-rounded-full;
      }

      &.--xl {
        @apply maz-px-8 maz-text-xl;

        padding-top: 1.325rem;
        padding-bottom: 1.325rem;
      }

      &.--lg {
        @apply maz-px-6 maz-text-lg;

        padding-top: 1rem;
        padding-bottom: 1rem;
      }

      &.--md {
        @apply maz-px-4;

        padding-top: 0.88rem;
        padding-bottom: 0.88rem;
      }

      &.--sm {
        @apply maz-px-3 maz-text-sm;

        padding-top: 0.625rem;
        padding-bottom: 0.625rem;
      }

      &.--xs {
        @apply maz-px-2 maz-text-sm;

        padding-top: 0.3rem;
        padding-bottom: 0.3rem;
      }

      &.--mini {
        @apply maz-px-1 maz-text-xs;

        padding-top: 0.2rem;
        padding-bottom: 0.2rem;
      }

      transition:
        background 300ms ease-in-out 0ms,
        color 300ms ease-in-out 0ms;

      /* Not disabled */

      &:not(.--disabled) {
        &:hover,
        &:focus {
          @apply maz-bg-gray-200;
        }
      }

      /* Fab */

      &.--fab {
        @apply maz-flex maz-items-center
          maz-justify-center maz-rounded-full maz-px-0 maz-py-0;

        &:not(.--no-elevation) {
          @apply maz-elevation;
        }

        &.--xl {
          @apply maz-h-[4.125rem] maz-w-[4.125rem] maz-text-xl;
        }

        &.--lg {
          @apply maz-h-[3.375rem] maz-w-[3.375rem] maz-text-lg;
        }

        &.--md {
          @apply maz-h-12 maz-w-12;
        }

        &.--sm {
          @apply maz-h-[2.375rem] maz-w-[2.375rem] maz-text-sm;
        }

        &.--xs {
          @apply maz-h-[1.725rem] maz-w-[1.725rem] maz-text-sm;
        }

        &.--mini {
          @apply maz-h-[1.4rem] maz-w-[1.4rem] maz-text-xs;
        }
      }

      &.--block {
        @apply maz-w-full;
      }

      &.--primary {
        @apply maz-bg-primary maz-text-primary-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-primary-600;
        }
      }

      &.--secondary {
        @apply maz-bg-secondary maz-text-secondary-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-secondary-600;
        }
      }

      &.--info {
        @apply maz-bg-info maz-text-info-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-info-600;
        }
      }

      &.--success {
        @apply maz-bg-success maz-text-success-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-success-600;
        }
      }

      &.--warning {
        @apply maz-bg-warning maz-text-warning-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-warning-600;
        }
      }

      &.--danger {
        @apply maz-bg-danger maz-text-white;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-danger-600;
        }
      }

      &.--white {
        @apply maz-bg-white maz-text-white-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-gray-300;
        }
      }

      &.--black {
        @apply maz-bg-black maz-text-black-contrast;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-gray-800;
        }
      }

      &.--transparent {
        @apply maz-bg-transparent;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-[#000]/[0.08] dark:maz-bg-[#FFF]/[0.08];
        }
      }

      &.--theme {
        @apply maz-bg-theme maz-text-color;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-theme-hover;
        }
      }

      /* OUTLINE */

      &.--primary-outline {
        @apply maz-border-primary maz-bg-primary-alpha-05 maz-text-primary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-primary maz-text-primary-contrast;
        }
      }

      &.--secondary-outline {
        @apply maz-border-secondary maz-bg-secondary-alpha-05 maz-text-secondary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-secondary maz-text-white;
        }
      }

      &.--info-outline {
        @apply maz-border-info maz-bg-info-alpha-05 maz-text-info;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-info maz-text-white;
        }
      }

      &.--success-outline {
        @apply maz-border-success maz-bg-success-alpha-05 maz-text-success;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-success maz-text-white;
        }
      }

      &.--danger-outline {
        @apply maz-border-danger maz-bg-danger-alpha-05 maz-text-danger;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-danger maz-text-white;
        }
      }

      &.--warning-outline {
        @apply maz-border-warning maz-bg-warning-alpha-05 maz-text-warning;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-warning maz-text-white;
        }
      }

      &.--white-outline {
        @apply maz-border-white maz-text-white;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-white maz-text-white-contrast;
        }
      }

      &.--black-outline {
        @apply maz-border-black maz-text-black;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-black maz-text-black-contrast;
        }
      }

      &.--theme-outline {
        @apply maz-border-theme maz-text-theme;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-theme maz-text-color;
        }
      }

      /* PASTEL */

      &.--primary-pastel {
        @apply maz-bg-primary-50 maz-text-primary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-primary maz-text-primary-contrast;
        }
      }

      &.--secondary-pastel {
        @apply maz-bg-secondary-50 maz-text-secondary;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-secondary maz-text-secondary-contrast;
        }
      }

      &.--info-pastel {
        @apply maz-bg-info-50 maz-text-info;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-info maz-text-info-contrast;
        }
      }

      &.--success-pastel {
        @apply maz-bg-success-50 maz-text-success;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-success maz-text-success-contrast;
        }
      }

      &.--danger-pastel {
        @apply maz-bg-danger-50 maz-text-danger;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-danger maz-text-danger-contrast;
        }
      }

      &.--warning-pastel {
        @apply maz-bg-warning-50 maz-text-warning;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-warning maz-text-warning-contrast;
        }
      }

      &.--white-pastel {
        @apply maz-bg-gray-100 maz-text-white;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-white maz-text-white-contrast;
        }
      }

      &.--black-pastel {
        @apply maz-bg-gray-200 maz-text-black;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-black maz-text-black-contrast;
        }
      }

      &.--theme-pastel {
        @apply maz-bg-gray-200 maz-text-black;

        &:not(:disabled):hover,
        &:not(:disabled):focus {
          @apply maz-bg-black maz-text-black-contrast;
        }
      }

      /* DISABLED */

      &.--disabled {
        @apply maz-cursor-not-allowed maz-bg-color-lighter maz-text-gray-400;
      }

      &.--loading {
        @apply maz-cursor-wait;
      }

      &.--no-padding {
        @apply maz-p-0;
      }
    }
  }
</style>
