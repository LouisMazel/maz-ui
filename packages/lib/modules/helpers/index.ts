export * from './inject-strict'
export * from './debounce'
export * from './truthy-filter'
export * from './mount-component'
export * from './idle-timeout'
export * from './user-visibility'
export * from './sleep'
export * from './is-client'
export * from './script-loader'
export * from './country-code-to-unicode-flag'
export * from './swipe-handler'
export * from './throttle'
