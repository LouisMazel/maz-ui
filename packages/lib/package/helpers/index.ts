export { injectStrict } from './inject-strict'
export { debounce } from './debounce'
export { truthyFilter } from './truhty-filter'
export { mount } from './mount-component'
