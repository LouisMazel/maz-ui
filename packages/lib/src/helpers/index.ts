export * from './checkAvailability'
export * from './countryCodeToUnicodeFlag'
export * from './countryFlagFromFlagCdn'
export * from './debounce'
export * from './debounceCallback'
export * from './debounceId'
export * from './freezeValue'
export * from './idleTimeout'
export * from './injectStrict'
export * from './isClient'
export * from './isEqual'
export * from './isStandaloneMode'
export * from './mountComponent'
export * from './normalizeString'
export * from './scriptLoader'
export * from './sleep'
export * from './swipeHandler'
export * from './throttle'
export * from './throttleId'
export * from './truthyFilter'
export * from './userVisibility'
