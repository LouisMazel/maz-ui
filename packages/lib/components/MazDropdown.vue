<template>
  <div :id="instanceId" class="m-dropdown">
    <MazBtn
      :color="color"
      size="md"
      :aria-expanded="dropdownOpen"
      aria-haspopup="menu"
      :disabled="disabled"
      v-bind="$attrs"
      @click.stop="['click'].includes(trigger) ? toggleDropdown() : undefined"
      @focus="['hover', 'both'].includes(trigger) ? setDropdown(true) : undefined"
      @blur="setDropdownDebounced(false)"
      @keydown="keyboardOpenDropdown"
      @mouseenter="
        ['hover', 'both'].includes(trigger)
          ? dropdownOpen === false
            ? setDropdown(true)
            : setDropdownDebounced(true)
          : undefined
      "
      @mouseleave="['hover', 'both'].includes(trigger) ? setDropdownDebounced(false) : undefined"
    >
      <!-- @slot Menu Label -->
      <slot></slot>
    </MazBtn>

    <Transition name="maz-scale-fade">
      <div
        v-if="dropdownOpen"
        role="menu"
        aria-label="Menu"
        class="menu"
        tabindex="-1"
        :class="{
          '--top': position.includes('top'),
          '--left': position.includes('left'),
          '--right': position.includes('right'),
          '--bottom': position.includes('bottom'),
        }"
        @focus="setDropdownDebounced(true)"
        @blur="setDropdownDebounced(false)"
        @mouseenter="['hover', 'both'].includes(trigger) ? setDropdownDebounced(true) : undefined"
        @mouseleave="['hover', 'both'].includes(trigger) ? setDropdownDebounced(false) : undefined"
      >
        <template v-for="(item, index) in items">
          <!--
            @slot Custom menu item
              @binding {Object} item - menu item
          -->
          <slot name="menuitem" :item="item">
            <Component
              :is="item.to ? 'router-link' : item.action ? 'button' : 'a'"
              :key="index"
              :href="item.href"
              tabindex="-1"
              :target="item.href ? item.target ?? '_self' : undefined"
              :type="item.action ? 'button' : undefined"
              :class="[
                {
                  '--is-keyboard-selected': keyboardSelectedIndex === index,
                },
                item.class,
              ]"
              :to="item.to"
              class="menuitem"
              @click.stop="item.action ? runAction(item, $event) : closeOnClick()"
            >
              <!--
                @slot Custom menu item label
                  @binding {Object} item - menu item
              -->
              <slot name="menuitem-label" :item="item">
                {{ item.label }}
              </slot>
            </Component>
          </slot>
        </template>
      </div>
    </Transition>
  </div>
</template>

<script lang="ts" setup>
  import { ref, watch, getCurrentInstance } from 'vue'
  import MazBtn, { type Color } from './MazBtn.vue'
  import { useInstanceUniqId } from '../modules/composables/use-instance-uniq-id'
  import { type RouteLocationRaw } from 'vue-router'
  import { debounce } from '../modules/helpers/debounce'
  import { type Position } from './types'

  export type { Color, Position }

  export type MenuItem = {
    label: string
    action?: () => unknown
    target?: string
    href?: string
    to?: RouteLocationRaw
    class?: string
  } & Record<string, unknown>

  const props = withDefaults(
    defineProps<{
      /** Menu items */
      items: MenuItem[]
      /** Menu should be open? */
      open?: boolean
      /** id of the menu */
      id?: string
      /** Should open the dropdown on click, hover or both */
      trigger?: 'click' | 'hover' | 'both'
      /** Button color */
      color?: Color
      /** Position of the dropdown */
      position?: Position
      /** Disable close menu on menuitem clicked */
      noCloseOnClick?: boolean
      /** Disable menu */
      disabled?: boolean
    }>(),
    {
      id: 'maz-dropdown',
      trigger: 'both',
      color: 'transparent',
      position: 'bottom left',
    },
  )

  const emits = defineEmits([
    /**
     * emitted when a menu item is clicked
     */
    'menuitem-clicked',
    /**
     * Triggers when the number changes
     *
     * @property {boolean} open new value
     */
    'update:open',
  ])

  const instanceId = useInstanceUniqId({
    componentName: 'MazCheckbox',
    instance: getCurrentInstance(),
    providedId: props.id,
  })

  const dropdownOpen = ref(props.open)
  const keyboardSelectedIndex = ref<number>()

  function toggleDropdown() {
    setDropdown(!dropdownOpen.value)
  }

  const setDropdown = (value: boolean) => {
    if (props.disabled) return

    dropdownOpen.value = value
    emits('update:open', value)
  }
  const setDropdownDebounced = debounce((value: boolean) => {
    setDropdown(value)
  }, 200)

  async function runAction(item: MenuItem, event: Event) {
    emits('menuitem-clicked', event)

    await item.action?.()

    closeOnClick()
  }

  function closeOnClick() {
    if (props.noCloseOnClick === false) setDropdown(false)
  }

  function keydownHandler(event: KeyboardEvent) {
    if (event.key === 'Escape') {
      event.preventDefault()
      setDropdown(false)
    } else if (['ArrowDown', 'ArrowUp'].includes(event.key)) {
      arrowHandler(event)
    } else if (event.key === 'Enter' && typeof keyboardSelectedIndex.value === 'number') {
      event.preventDefault()
      const item = document.querySelectorAll(`#${instanceId.value} .menuitem`)[
        keyboardSelectedIndex.value
      ] as HTMLElement

      item.click()

      closeOnClick()
    }
  }

  function keyboardOpenDropdown(event: KeyboardEvent) {
    if (['ArrowDown', 'ArrowUp', 'Enter'].includes(event.key) && dropdownOpen.value === false) {
      event.preventDefault()
      setDropdown(true)
    }
  }

  function arrowHandler(event: KeyboardEvent) {
    event.preventDefault()
    const code = event.key

    if (!dropdownOpen.value) setDropdown(true)

    const itemLength = props.items?.length

    if (!itemLength) return

    if (typeof keyboardSelectedIndex.value === 'number') {
      if (keyboardSelectedIndex.value === itemLength - 1 && code === 'ArrowDown') {
        keyboardSelectedIndex.value = 0
      } else if (keyboardSelectedIndex.value === 0 && code === 'ArrowUp') {
        keyboardSelectedIndex.value = itemLength - 1
      } else {
        keyboardSelectedIndex.value =
          code === 'ArrowDown' ? keyboardSelectedIndex.value + 1 : keyboardSelectedIndex.value - 1
      }
    } else {
      keyboardSelectedIndex.value = code === 'ArrowDown' ? 0 : itemLength - 1
    }
  }

  watch(
    () => dropdownOpen.value,
    (value) => {
      if (value) {
        document.addEventListener('keydown', keydownHandler)
      } else {
        document.removeEventListener('keydown', keydownHandler)
      }

      keyboardSelectedIndex.value = undefined
    },
  )
  watch(
    () => props.open,
    (value) => setDropdown(value),
  )
</script>

<style lang="postcss">
  .m-dropdown {
    @apply maz-relative maz-inline-flex maz-flex-col maz-items-start;

    & [aria-expanded='true'].m-btn {
      @apply maz-bg-color-light;
    }

    .menu {
      @apply maz-absolute maz-z-default-backdrop maz-flex maz-flex-col maz-gap-0.5 maz-overflow-auto maz-rounded maz-bg-color maz-p-2 maz-elevation dark:maz-border dark:maz-border-color-light;

      &.--top:not(.--right, .--left) {
        @apply maz-bottom-full maz-mb-1 maz-origin-bottom;
      }

      &.--bottom:not(.--right, .--left) {
        @apply maz-top-full maz-mt-1 maz-origin-top;
      }

      &.--left:not(.--top, .--bottom) {
        @apply maz-right-full maz-mr-1 maz-origin-top-right;
      }

      &.--right:not(.--top, .--bottom) {
        @apply maz-left-full maz-ml-1 maz-origin-top-left;
      }

      &.--top.--right {
        @apply maz-bottom-full maz-right-0 maz-mb-1 maz-origin-bottom-right;
      }

      &.--top.--left {
        @apply maz-bottom-full maz-left-0 maz-mb-1 maz-origin-bottom-left;
      }

      &.--bottom.--left {
        @apply maz-left-0 maz-top-full maz-mt-1 maz-origin-top-left;
      }

      &.--bottom.--right {
        @apply maz-right-0 maz-top-full maz-mt-1 maz-origin-top-right;
      }

      .menuitem {
        @apply maz-cursor-pointer maz-whitespace-nowrap maz-rounded maz-px-4 maz-py-2 maz-text-left maz-outline-none maz-transition-colors maz-duration-300 maz-ease-in-out focus-within:maz-bg-color-light hover:maz-bg-color-light;

        &.--is-keyboard-selected {
          @apply maz-bg-color-light;
        }
      }
    }
  }
</style>
