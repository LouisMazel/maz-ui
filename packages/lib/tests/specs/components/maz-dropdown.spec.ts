import { mount } from '@vue/test-utils'
import MazDropdown from '@components/MazDropdown.vue'
import MazBtn from '@components/MazBtn.vue'
import { sleep } from '@modules/helpers'

describe('MazDropdown', () => {
  test('renders a dropdown button with the provided label', () => {
    const wrapper = mount(MazDropdown, {
      slots: {
        default: 'Dropdown Label',
      },
      props: {
        items: [{ label: 'Menu 1', action: vi.fn() }],
      },
    })

    const dropdownButton = wrapper.findComponent(MazBtn)
    expect(dropdownButton.exists()).toBe(true)
    expect(dropdownButton.text()).toBe('Dropdown Label')
  })

  test('toggles the dropdown when the button is focused', async () => {
    const wrapper = mount(MazDropdown, {
      slots: {
        default: 'Dropdown Label',
      },
      props: {
        trigger: 'click',
        items: [{ label: 'Menu 1', action: vi.fn() }],
      },
    })

    const dropdownButton = wrapper.findComponent(MazBtn)

    await wrapper.vm.$nextTick()

    expect(wrapper.vm.dropdownOpen).toBe(false)

    await dropdownButton.trigger('click')
    expect(wrapper.vm.dropdownOpen).toBe(true)

    await dropdownButton.trigger('click')
    expect(wrapper.vm.dropdownOpen).toBe(false)
  })

  test('toggles the dropdown when the button is clicked', async () => {
    const wrapper = mount(MazDropdown, {
      slots: {
        default: 'Dropdown Label',
      },
      props: {
        trigger: 'both',
        items: [{ label: 'Menu 1', action: vi.fn() }],
      },
    })

    const dropdownButton = wrapper.findComponent(MazBtn)

    await wrapper.vm.$nextTick()

    expect(wrapper.vm.dropdownOpen).toBe(false)

    await dropdownButton.trigger('focus')
    expect(wrapper.vm.dropdownOpen).toBe(true)

    await dropdownButton.trigger('blur')
    await sleep(200)
    expect(wrapper.vm.dropdownOpen).toBe(false)
  })

  test('emits "update:open" event when the dropdown is toggled', async () => {
    const wrapper = mount(MazDropdown, {
      slots: {
        default: 'Dropdown Label',
      },
      props: {
        items: [{ label: 'Menu 1', action: vi.fn() }],
      },
    })

    const dropdownButton = wrapper.findComponent(MazBtn)

    await dropdownButton.trigger('focus')

    expect(wrapper.emitted('update:open')).toBeTruthy()
    expect(wrapper.emitted('update:open')[0]).toEqual([true])

    await dropdownButton.trigger('blur')

    await sleep(200)

    expect(wrapper.emitted('update:open')).toBeTruthy()
    expect(wrapper.emitted('update:open')[1]).toEqual([false])
  })

  test('closes the dropdown when a menu item is clicked', async () => {
    const action = vi.fn()
    const wrapper = mount(MazDropdown, {
      slots: {
        default: 'Dropdown Label',
      },
      props: {
        items: [{ label: 'Menu 1', action }],
      },
    })

    const dropdownButton = wrapper.findComponent(MazBtn)

    await dropdownButton.trigger('focus')
    expect(wrapper.vm.dropdownOpen).toBe(true)

    const menuItems = wrapper.findAll('.menuitem')
    await menuItems[0].trigger('click')

    expect(action).toHaveBeenCalledTimes(1)

    await sleep(200)

    expect(wrapper.vm.dropdownOpen).toBe(false)
  })

  test('navigates through menu items using the keyboard', async () => {
    const wrapper = mount(MazDropdown, {
      slots: {
        default: 'Dropdown Label',
        menuitem: '<div class="menuitem">Menu Item 1</div><div class="menuitem">Menu Item 2</div>',
      },
      props: {
        items: [
          { label: 'Menu 1', action: vi.fn() },
          { label: 'Menu 2', action: vi.fn() },
        ],
      },
    })

    const dropdownButton = wrapper.findComponent(MazBtn)

    await dropdownButton.trigger('focus')
    expect(wrapper.vm.dropdownOpen).toBe(true)

    // Simulate ArrowDown key
    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowDown' }))
    expect(wrapper.vm.keyboardSelectedIndex).toBe(0)

    // Simulate ArrowUp key
    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowDown' }))
    expect(wrapper.vm.keyboardSelectedIndex).toBe(1)

    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowDown' }))
    expect(wrapper.vm.keyboardSelectedIndex).toBe(0)
  })
})
