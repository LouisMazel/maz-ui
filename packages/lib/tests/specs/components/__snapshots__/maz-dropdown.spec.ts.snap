// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MazDropdown > Should match with the snapshot 1`] = `
"<div id=\\"maz-dropdown\\" class=\\"m-dropdown\\"><button data-v-fa7305fb=\\"\\" aria-expanded=\\"false\\" aria-haspopup=\\"menu\\" class=\\"m-btn --md --transparent --cursor-pointer --is-button\\" type=\\"button\\">
    <!--v-if-->
    <!--v-if--><span data-v-fa7305fb=\\"\\" class=\\"maz-flex maz-flex-center\\"></span>
    <!--v-if-->
    <!--v-if-->
  </button>
  <transition-stub name=\\"maz-scale-fade\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
    <!--v-if-->
  </transition-stub>
</div>"
`;
